x/twap/listeners_test.go

```diff
@@ -77,58 +78,222 @@ func (s *TestSuite) TestAfterPoolCreatedHook() {
+			s.Require().NoError(err)
+
+			// if no swap happened in block1, there should be no change
+			// in the most recent twap record after epoch
+			if !tc.block1Swap {
+				s.Require().Equal(twapAfterPoolCreation, twapAfterBlock1)
+			} else {
+				// height should not have changed
+				s.Require().Equal(twapAfterPoolCreation.Height, twapAfterBlock1.Height)
+				// twap time should be same as previous blocktime
+				s.Require().Equal(twapAfterPoolCreation.Time, baseTime)
+
+				// accumulators should not have increased, as they are going through the first epoch
+				s.Require().Equal(sdk.ZeroDec(), twapAfterBlock1.P0ArithmeticTwapAccumulator)
+				s.Require().Equal(sdk.ZeroDec(), twapAfterBlock1.P1ArithmeticTwapAccumulator)
+			}
+
+			// check if spot price has been correctly updated in twap record
+			asset0sp, err := s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapAfterBlock1.Asset0Denom, twapAfterBlock1.Asset1Denom)
+			s.Require().NoError(err)
+			s.Require().Equal(asset0sp, twapAfterBlock1.P0LastSpotPrice)
+
+			// run basic swap on block two for price change
+			if tc.block2Swap {
+				s.RunBasicSwap(poolId)
+			}
 
-	s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+			s.EndBlock()
+			s.Commit()
+
+			// Third block
+			twapAfterBlock2, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
+			s.Require().NoError(err)
+
+			// if no swap happened in block 3, twap record should be same with block 2
+			if !tc.block2Swap {
+				s.Require().Equal(twapAfterBlock1, twapAfterBlock2)
+			} else {
+				s.Require().Equal(secondBlockTime, twapAfterBlock2.Time)
+
+				// check accumulators incremented - we test details of correct increment in logic
+				s.Require().True(twapAfterBlock2.P0ArithmeticTwapAccumulator.GT(twapAfterBlock1.P0ArithmeticTwapAccumulator))
+				s.Require().True(twapAfterBlock2.P1ArithmeticTwapAccumulator.GT(twapAfterBlock1.P1ArithmeticTwapAccumulator))
+			}
+
+			// check if spot price has been correctly updated in twap record
+			asset0sp, err = s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapAfterBlock1.Asset0Denom, twapAfterBlock2.Asset1Denom)
+			s.Require().NoError(err)
+			s.Require().Equal(asset0sp, twapAfterBlock2.P0LastSpotPrice)
+		})
+	}
 }
 
-// TestSwapAndEndBlockTriggeringSave tests that if we:
-// * create a pool in block 1
-// * swap in block 2
-// then after block 2 end block, we have saved records for the pool,
-// for both block 1 & 2, with distinct spot prices in their records, and accumulators incremented.
-// TODO: Abstract this to be more table driven, and test more pool / block setups.
-func (s *TestSuite) TestSwapAndEndBlockTriggeringSave() {
+// TestAfterEpochEnd tests if records get succesfully deleted via `AfterEpochEnd` hook.
+// We test details of correct implementation of pruning method in store test.
+// TODO: improve test using `GetAllHistoricalTimeIndexedTWAPs` and `GetAllHistoricalPoolIndexedTWAPs`
```

Comment:
Do you mind making an issue for this please outlining specifically how this can be improved?

Reply:
added https://github.com/osmosis-labs/osmosis/issues/2431 here!